framework/src/main/java/org/checkerframework/framework/source/SourceChecker.java

```diff
@@ -1079,6 +1084,43 @@ protected void printOrStoreMessage(
         Trees.instance(processingEnv).printMessage(kind, message, source, root);
     }
 
+    /**
+     * printOrStoreMessage method with an added stack trace argument. The stack trace is printed
+     * when the dumpOnErrors option is enabled.
+     *
+     * @param kind the kind of message to print
+     * @param message the message text
+     * @param source the souce code position of the diagnostic message
+     * @param root the compilation unit
+     * @param trace the stack trace where the checker encountered an error
+     */
+    protected void printOrStoreMessage(
+            javax.tools.Diagnostic.Kind kind,
+            String message,
+            Tree source,
+            CompilationUnitTree root,
+            StackTraceElement[] trace) {
+        Trees.instance(processingEnv).printMessage(kind, message, source, root);
+        printStackTrace(trace);
+    }
+
+    /**
+     * Output the given stack trace if the "dumpOnErrors" option is enabled.
+     *
+     * @param trace stack trace when the checker encountered a warning/error
+     */
+    private void printStackTrace(StackTraceElement[] trace) {
+        boolean dumpOnErrors =
+                getOptions().containsKey("dumpOnErrors") && getBooleanOption("dumpOnErrors", true);
+        if (dumpOnErrors) {
+            String msg = new String();
```

Comment:
Don't build up with just a String! In particular that initial `new String()` is very ugly.
Use a StringBuilder or a StringJoiner.

Reply:
I have used a StringBuilder to build the stack trace and print it as a note in my latest commit.